### PR TITLE
✨ Adding custom queue

### PR DESCRIPTION
**Pull request by YousefEZ**
- ✨ Added FRRQueue that takes in two template parameters, the Congestion Policy, and the FRR Policy
  - the Congestion Policy determines when to mark the queue as congested
  - the FRR Policy determines how to select the alternative target
- ✨ Added Module Congestion Policy (Every other enqueue, mark as failed)
- ✨ Added LFA FRR Policy
- ✨ Added script to test changes, with layout into how to easily test different policies

### CODE REVIEW

File: libs/dummy_congestion_policy.h
Comment (charliebarber) [inline libs/dummy_congestion_policy.h:1]: Looks good, query raised on isCongested privately.

File: libs/lfa_policy.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+class LFAPolicy
+{
+  private:  
+    std::list<ns3::Ptr<ns3::PointToPointNetDevice>> m_alternateTargets;
+
+  public:
+    LFAPolicy() = default;
+ 
+    template <typename... DEVICES>
+    void addAlternateTargets(DEVICES&&... devices);
```
Comment (charliebarber) [inline libs/lfa_policy.h:23]: hadn't seen this syntax before but makes sense. all very efficient with the forwarding going on so good stuff

File: libs/modulo_congestion_policy.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  private:
+    int counter; 
+
+  public:
+
+    template <typename CONTAINER>
+    bool isCongested(const CONTAINER& container)
+    {
+      increment();
+      return !!counter;
```
Comment (charliebarber) [inline libs/modulo_congestion_policy.h]: To be honest I would lean towards counter != 0 if that's what you're trying to achieve. Just for readability but fine either way

File: src/lfa.cc [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    address.SetBase("10.1.1.0", "255.255.255.0");
+    Ipv4InterfaceContainer interfaces01 = address.Assign(devices01);
+    address.SetBase("10.1.2.0", "255.255.255.0");
+    Ipv4InterfaceContainer interfaces12 = address.Assign(devices12);
+    address.SetBase("10.1.3.0", "255.255.255.0");
+    Ipv4InterfaceContainer interfaces02 = address.Assign(devices02);
+
+    Ipv4GlobalRoutingHelper::PopulateRoutingTables();
+
+    // Configure the application to generate traffic
```
Comment (charliebarber) [inline src/lfa.cc:89]: This isn't quite what i thought we had envisioned for simulating congestion but it works. Don't worry about that part however.

**Review by charliebarber — APPROVED** [review]
Really well written C++ code and everything makes sense. I agree with all your design choices and queues in this manner is definitely the best approach to handling FRR. Good stuff.

File: libs/modulo_congestion_policy.h (hunk 2 above)
Comment (YousefEZ) [reply libs/modulo_congestion_policy.h]: tbh can just return counter and it'll convert it implicitly

File: src/lfa.cc (hunk 3 above)
Comment (YousefEZ) [reply src/lfa.cc:89]: the sample file is for testing that we can reroute based on CongestionPolicy and using the FRRPolicy more than simulating congestion tbh

**Review by charliebarber — APPROVED** [review]

**Pull request merged**